[libcurl-simple-https] New port

SamuelMarks opened the pull request:
**Describe the pull request**
>  Very simple HTTPS interface built atop libcurl 

https://github.com/SamuelMarks/curl-simple-https

- #### What does your PR fix?  

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  

All

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  

Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  

Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

Possibly still a draft, need to see what your CI says

SamuelMarks committed:
[ports/libcurl-simple-https] Init

SamuelMarks committed:
[versions/d-/dmon.json,versions/baseline.json] `./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] Add file forgotten from `git add`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Upgrade (uses latest FindCurl impl)

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Upgrade to boilerplate conforming version

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Upgrade to latest libcurl-simple-https

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https] Upgrade to latest version (which actually exports targets… and renames to dashes not underscore)

SamuelMarks committed:
[versions/baseline.json,versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version` + manually remove yesterdays

SamuelMarks commented:
Actually don't merge this yet, getting weirdness on the `find_package` end. Without `find_package` it links fine with my package ([cauthflow](https://github.com/offscale/cauthflow/runs/5078410254)). With `find_package(libcurl-simple-https CONFIG REQUIRED)` I get:

```
Could not find a package configuration file provided by
  "libcurl-simple-https" with any of the following names:

    libcurl-simple-httpsConfig.cmake
    libcurl-simple-https-config.cmake
```

Even though:
```
vcpkg (project0) $ fd -HIFecmake 'libcurl-simple-httpsConfig'
buildtrees/libcurl-simple-https/x64-osx-dbg/libcurl-simple-httpsConfig.cmake
buildtrees/libcurl-simple-https/x64-osx-dbg/libcurl-simple-httpsConfigVersion.cmake
buildtrees/libcurl-simple-https/x64-osx-rel/libcurl-simple-httpsConfig.cmake
buildtrees/libcurl-simple-https/x64-osx-rel/libcurl-simple-httpsConfigVersion.cmake
```

(and it found my other vcpkg dependencies, it's just this package…)

dg0yt commented:
`set(CMAKE_FIND_DEBUG_MODE 1)`

SamuelMarks commented:
@dg0yt Ok thanks… so with that enabled I was able to find that manifest mode is the issue. I have a `vcpkg.json` in my root.

With manifest mode enables I still get this:

```
The package libcurl-simple-https provides CMake targets:

    find_package(libcurl-simple-https CONFIG REQUIRED)
    target_link_libraries(main PRIVATE libcurl-simple-https libcurl-simple-https_compiler_flags)
```

(even double checked by `./vcpkg remove libcurl-simple-https` first from my vcpkg root)

However `fd -HIFecmake 'libcurl-simple-httpsConfig'` shows nothing from my project root (only shows in my vcpkg root when I `./vcpkg install libcurl-simple-https`)

What am I doing wrong? - E.g., [this line of my `CMakeLists.txt`](https://github.com/SamuelMarks/curl-simple-https/blob/61cc5c4/libcurl-simple-https/CMakeLists.txt#L58):
```cmake
install(EXPORT "${LIBRARY_NAME}Targets" DESTINATION "share/${LIBRARY_NAME}")
```

dg0yt commented:
I am unable to follow. Maybe it is too early to integrate into vcpkg.

SamuelMarks commented:
@dg0yt Possibly. Anyway this project is the simplest non-header-only C89 CMake project I have. So once this issue is resolved then you'll see a lot more [& working] ports being contributed from me.

There are two `CMakeLists.txt` one .c and one .h, along with a `vcpkg.json`:
```
|-- CMakeLists.txt
|-- libcurl-simple-https
|   |-- CMakeLists.txt
|   |-- curl_simple_https.c
|   `-- curl_simple_https.h
`-- vcpkg.json
```

Everything else is boilerplate, primarily from the official CMake tutorial.

Why is this failing? - Thanks for any insight

dg0yt commented:
> Anyway this project is the simplest non-header-only C89 CMake project I have. So once this issue is resolved then you'll see a lot more [& working] ports being contributed from me.

C89 seems to be quite the opposite direction to modern C++. I have no benefit from the fact that a package supports this old standard.
And ports which are working now are still a technical debt when updating their dependencies. (And I did work on port curl.)

> Everything else is boilerplate, primarily from the official CMake tutorial.

No, it isn't. It does a lot of non-standard things. (Hint: Many relevant lines contain `"${PROJECT`). To be fair, many CMake projects are more complex than necessary. 

> along with a vcpkg.json

Assuming that we talk about building the port, remove it (with a command in the portfile). 
Check the logs for what is build and installed, and where to.

SamuelMarks commented:
For reuse as boilerplate across my decent number of C projects, I took https://gitlab.kitware.com/cmake/cmake/-/blob/master/Help/guide/tutorial/Complete and:

  - replaced literal names with variables;
  - removed functionality irrelevant to my purposes;
  - added stronger `use strict` style guards:
    - [`set(CMAKE_C_STANDARD 90)`](https://cmake.org/cmake/help/latest/variable/CMAKE_C_STANDARD.html)
    - [`set(CMAKE_C_VISIBILITY_PRESET hidden)`](https://cmake.org/cmake/help/latest/prop_tgt/LANG_VISIBILITY_PRESET.html)
    - [`set(CMAKE_VISIBILITY_INLINES_HIDDEN YES)`](https://cmake.org/cmake/help/latest/variable/CMAKE_VISIBILITY_INLINES_HIDDEN.html)
    - `"$<${gcc_like}:$<BUILD_INTERFACE:-Wall;-Wextra;-Wshadow;-Wformat=2;-Wunused;-pedantic>>"`
    - `"$<${msvc}:$<BUILD_INTERFACE:-W3;-WX;-Zi;-permissive->>"`
  - generated export headers with [`GenerateExportHeader`](https://cmake.org/cmake/help/latest/module/GenerateExportHeader.html);
  - [`set`](https://cmake.org/cmake/help/latest/command/set.html) separate `Source_Files` and `Header_Files` vars [with [`source_groups`](https://cmake.org/cmake/help/latest/command/source_group.html)] for installation [and IDE assist]

Following [your advice](https://github.com/microsoft/vcpkg/pull/22930#discussion_r799588044) @dg0yt from the #22930 pull request I added the aforementioned [`INSTALL(EXPORT`](https://cmake.org/cmake/help/latest/command/install.html#export) line.

I'm going to try without manifest mode and see how far I get…

Any other tips would be greatly appreciated

SamuelMarks committed:
[ports/libcurl-simple-https] Upgrade to latest version (renames "_" to "-"; installs config file

SamuelMarks committed:
[versions/baseline.json,versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version` + manually remove yesterdays

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] `configure_package_config_file` with `INSTALL_DESTINATION` from `"lib/cmake/${PROJECT_NAME}"` to `"share/${PROJECT_NAME}"`

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] ./vcpkg x-add-version --all --overwrite-version

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Manual `install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake" DESTINATION "share/${LIBRARY_NAME}")` because `configure_package_config_file` `INSTALL_DESTINATION` does not seem to actually `install`…

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] ./vcpkg x-add-version --all --overwrite-version

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] `install` ""${PROJECT_NAME}Config.cmake" to "share/${PROJECT_NAME}/${LIBRARY_NAME}"

SamuelMarks committed:
[versions/l-/libcurl-simple-https.json] ./vcpkg x-add-version --all --overwrite-version

SamuelMarks commented:
FYI: It seems that I've gotten it to work now (haven't investigated why some of your CI isn't passing… will revisit tomorrow)

JonLiu1993 commented:
@SamuelMarks , Thanks for your pr, It looks like this error is causing many triples to fail on ci:
````
The following dlls were found in /lib or /debug/lib. Please move them to /bin or /debug/bin, respectively.

    D:/packages/libcurl-simple-https_x86-windows/lib/libcurl-simple-https.dll

Mismatching number of debug and release binaries. Found 0 for debug but 1 for release.
Debug binaries


Release binaries

    D:/packages/libcurl-simple-https_x86-windows/lib/libcurl-simple-https.lib

Debug binaries were not found

Found 2 post-build check problem(s). To submit these ports to curated catalogs, please first correct the portfile:
    C:\a\1\s\ports\libcurl-simple-https\portfile.cmake
-- Performing post-build validation done
Error: Building package libcurl-simple-https:x86-windows failed with: POST_BUILD_CHECKS_FAILED
Elapsed time for package libcurl-simple-https:x86-windows: 3.008 s

````

SamuelMarks commented:
@JonLiu1993 Yeah I'm getting the same issue on cauthflow. Been experimenting:
```cmake
# control where the static and shared libraries are built so that on windows
# we don't need to tinker with the path to run the executable
set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")
set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")
set(CMAKE_RUNTIME_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}")

if (NOT DEFINED CMAKE_BUILD_TYPE)
    if (DEFINED CMAKE_RC_FLAGS_RELEASE)
        set(CMAKE_BUILD_TYPE "Release")
    else ()
        set(CMAKE_BUILD_TYPE "Debug")
    endif ()
endif (NOT DEFINED CMAKE_BUILD_TYPE)

if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
    set(INSTALL_LIB_DESTINATION "bin")
    if (CMAKE_BUILD_TYPE STREQUAL "Debug")
        set(INSTALL_LIB_DESTINATION "${CMAKE_BUILD_TYPE}/${INSTALL_LIB_DESTINATION}")
    endif (CMAKE_BUILD_TYPE STREQUAL "Debug")
else()
    set(INSTALL_LIB_DESTINATION "lib")
endif (CMAKE_SYSTEM_NAME STREQUAL "Windows")
```

Later:
```cmake
set(installable_libs
        "${LIBRARY_NAME}" "${PROJECT_NAME}_compiler_flags" "${PROJECT_NAME}_utils" "${PROJECT_NAME}_google_auth")
if (TARGET "${DEPENDANT_LIBRARY}")
  list(APPEND installable_libs "${DEPENDANT_LIBRARY}")
endif (TARGET "${DEPENDANT_LIBRARY}")
install(TARGETS ${installable_libs}
        DESTINATION "${INSTALL_LIB_DESTINATION}"
        EXPORT "${LIBRARY_NAME}Targets")
install(FILES "${_export_file}" "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}_configure.h" ${Header_Files}
        DESTINATION "include")
install(EXPORT "${LIBRARY_NAME}Targets" DESTINATION "share/${LIBRARY_NAME}")
```

What am I doing wrong?

SamuelMarks committed:
[ports/libcurl-simple-https] Upgrade to latest version (which should support vcpkg install hierarchy)

SamuelMarks committed:
[versions/baseline.json,versions/l-/libcurl-simple-https.json] `./vcpkg x-add-version --all --overwrite-version` + manually remove old

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")`

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Use `ARCHIVE` and `LIBRARY` DESTINATION

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] Remove `file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug")`

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" ${CURRENT_PACKAGES_DIR}/debug/share")

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include" ${CURRENT_PACKAGES_DIR}/debug/share")

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] `install` `"${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}Config.cmake` to `DESTINATION` `"${CMAKE_INSTALL_DATADIR}/${PROJECT_NAME}"` not `TYPE "DATA"`

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

JonLiu1993 commented:
Trigger CLA check again

JonLiu1993 commented on ports/libcurl-simple-https/portfile.cmake:
```diff
@@ -0,0 +1,20 @@
+file(INSTALL "${SOURCE_PATH}/cmake/License.txt"
+     DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}"
+     RENAME copyright)
+foreach(_dir "include" "share")
+    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/${_dir}")
+endforeach(_dir "include" "share")
```
```suggestion
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/share")
    file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug/include")
```

JonLiu1993 replied:
@SamuelMarks ,Thanks for your pr, LGTM if you can If write this way

SamuelMarks replied:
@JonLiu1993 Ready for merge

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [CMakeLists.txt] Bump `cmake_minimum_required` from `3.0` to `3.3`; [libcurl-simple-https/curl_simple_https.c] Use `calloc`; add debug statements

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] No alt option set when body is NULL

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] More debug statements

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] More debug statements

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] More debug statements

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] Replace `assert` with `if` for `curl_url_get`

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

SamuelMarks committed:
[ports/libcurl-simple-https/portfile.cmake] [libcurl-simple-https/curl_simple_https.c] Set `CURLOPT_READDATA`, `CURLOPT_READFUNCTION`, `CURLOPT_POSTFIELDSIZE`: even with an empty HTTPS POST

SamuelMarks committed:
`./vcpkg x-add-version --all --overwrite-version`

dg0yt commented:
Personally I think packages should have some relevance *before* being accepted into the main repo.

sfhacker commented:
Perhaps someone has requested this addition.

SamuelMarks commented:
It's a convenient interface to libcurl?

Even this month this was posted https://daniel.haxx.se/blog/2022/02/02/curl-dash-dash-json

So it's not like they aren't thinking about it [officially]…

mathisloge commented:
> Personally I think packages should have some relevance _before_ being accepted into the main repo.

think so too. Such just-created projects are often abandoned by the maintainer and are then in the official repo and carry a lot of dept. Especially when you consider that everything always has to be built, projects that are possibly only used by one person or no one can disrupt an important library such as curl in the update process. IMHO those are good candidates for custom registries. https://github.com/microsoft/vcpkg/pull/17540#issuecomment-829608222

SamuelMarks commented:
First commit: Jul 23, 2021

Albeit not used by projects I'm not involved in, but this is a dependency for:

  - [cauthflow](https://github.com/offscale/cauthflow)
  - bunch of internal projects
  - [google-cloud-c](https://github.com/offscale/google-cloud-c)

If there was some better metric for acceptance, happy to follow. E.g, a certain level of test and documentation coverage, cross-platform support, &etc.

mathisloge commented:
This is just my personal opinion and not specific to this project. 
But all these projects you list belong to your own organisation and would work even if they were used in a custom registry, and this will not prevent anyone from using your projects either. and if, for example, you realise at some point that you no longer want to support them, that's no problem either, because they don't have to be maintained in the vcpkg universe.

But yeah. This is a decision of the vcpkg team.

SamuelMarks commented:
Yes you are right that these would all work just fine in a custom registry. But that's not my question. My question is what metrics do you use to decide betwixt the two?

The primary advantages of having them in vcpkg itself are: review by vcpkg staff; CI access; advertising (when I was looking for UI frameworks I used `grep` and `ls` on your `ports` tree to find them). So would be good not to lose these advantages.

Metrics which might be what you're looking for to decide whether to accept them into vcpkg are, from the top of my head:

  - Cross-platform support;
  - Cross-architecture support (esp. for Windows);
  - Open-source license (and which one);
  - Test coverage;
  - Documentation coverage.

Fluffier metrics like number of tweets, blogposts, GitHub stars and the like are all gameable… so would prefer to stick to more objective measures (assuming you'd be open to that).

mathisloge commented:
I think it would be good to have a metric for survivablity.
E.g. How likely is it, that the project would be maintained in the future. 
So:

* founded support of companies 
* number of contributors
* stars could very well be a guidance: if the maintainer sometime decide to not support the project, some of those stargazers might jump in. So the likelyhood of survival is a bit better 

Test coverage etc. are not that indicative IMHO since those need to be analyzed very closly to say anything about the quality of the coverage.

dg0yt commented:
> The primary advantages of having them in vcpkg itself are: review by vcpkg staff; CI access; advertising

I do understand the advantages on your side. But I also see that your PRs consume limited review capacity, even solving "upstream" (i.e. your) problems with plain CMake project which are entirely unrelated to vcpkg. In the meantime I wait days and weeks for reviews, have to re-explain basics, and re-resolve merge conflicts. So your benefits come at a price. 
And I all I want that there is a conciousness for these side effects.
